.gas-snapshot

```diff
@@ -1,14 +1,11 @@
-WakuRlnV2Test:test__IdCommitmentToMetadata__DoesntExist() (gas: 16726)
-WakuRlnV2Test:test__InvalidPaginationQuery__EndIndexGTcommitmentIndex() (gas: 18249)
-WakuRlnV2Test:test__InvalidPaginationQuery__StartIndexGTEndIndex() (gas: 16130)
-WakuRlnV2Test:test__InvalidRegistration__DuplicateIdCommitment() (gas: 99502)
-WakuRlnV2Test:test__InvalidRegistration__FullTree() (gas: 14328)
-WakuRlnV2Test:test__InvalidRegistration__InvalidIdCommitment__LargerThanField() (gas: 15229)
-WakuRlnV2Test:test__InvalidRegistration__InvalidIdCommitment__Zero() (gas: 14004)
-WakuRlnV2Test:test__InvalidRegistration__InvalidUserMessageLimit__LargerThanMax() (gas: 17703)
-WakuRlnV2Test:test__InvalidRegistration__InvalidUserMessageLimit__Zero() (gas: 14085)
-WakuRlnV2Test:test__Upgrade() (gas: 3791109)
-WakuRlnV2Test:test__ValidPaginationQuery(uint32) (runs: 1000, μ: 445155, ~: 159972)
-WakuRlnV2Test:test__ValidPaginationQuery__OneElement() (gas: 119773)
-WakuRlnV2Test:test__ValidRegistration(uint256,uint32) (runs: 1000, μ: 124408, ~: 124408)
-WakuRlnV2Test:test__ValidRegistration__kats() (gas: 96616)
\ No newline at end of file
+WakuRlnV2Test:test__IdCommitmentToMetadata__DoesntExist() (gas: 27478)
+WakuRlnV2Test:test__InsertionNormalOrder(uint32) (runs: 1001, μ: 1123469, ~: 494837)
+WakuRlnV2Test:test__InvalidPaginationQuery__EndIndexGTcommitmentIndex() (gas: 18261)
+WakuRlnV2Test:test__InvalidPaginationQuery__StartIndexGTEndIndex() (gas: 16108)
+WakuRlnV2Test:test__InvalidRegistration__DuplicateIdCommitment() (gas: 249370)
+WakuRlnV2Test:test__InvalidRegistration__InvalidIdCommitment__Zero() (gas: 12135)
+WakuRlnV2Test:test__Upgrade() (gas: 6728616)
+WakuRlnV2Test:test__ValidPaginationQuery(uint32) (runs: 1002, μ: 226428, ~: 52927)
+WakuRlnV2Test:test__ValidPaginationQuery__OneElement() (gas: 262351)
+WakuRlnV2Test:test__ValidRegistration(uint256,uint32) (runs: 1001, μ: 268741, ~: 268741)
```

Comment:
a valid registration seems to go from `124408` to `268741` of gas, which is x2. Can we optimize it?

>  It's worth mentioning that keeping track of the expired memberships means some gas expense due to having to maintain the linking for the memberships in this list

Ideally the contract should be as simple as possible, and I would say that's possible but we may have to circle back to the spec. Perhaps some things can be handled offchain, or well, its state does not need to be tracked onchain.

Reply:
I can try optimizing it, but the main problem IMO is the additional storage required due to https://github.com/waku-org/waku-rlnv2-contract/blob/251d8907d4d7ee71e77fdb9ae08e1d4cd32f9f23/src/Membership.sol#L39-L96

We could probably cut back the cost by not mainainting the ordered membership list, and have nwaku be the one tracking the age of the memberships and deciding then which memberships to erase. I'd actually prefer that since we make the contract cheaper for the end user, but I don't know if keeping track of the membership age in nwaku is something we want to do?

Reply:
> We could probably cut back the cost by not mainainting the ordered membership list, and have nwaku be the one tracking the age of the memberships and deciding then which memberships to erase. I'd actually prefer that since we make the contract cheaper for the end user, but I don't know if keeping track of the membership age in nwaku is something we want to do?

do you mean that nwaku would need to listen to event to know age?

Is the issue that the list is ordered on the contract (I am not sure which piece of code you are referring with "due to")?

Why ordering the list?

Reply:
> The new membership SHOULD overwrite the membership that has been Expired for the longest time

I think it would make sense to drop this spec requirement and move to a model where the caller can pass a membership index to overwrite.

Ideally, we should store expiry time (or registration time + length) on the smart contract so we don't need to listen to events.